### PR TITLE
stage 2 model implementation

### PR DESCRIPTION
**Pull request by Devin-Taylor**
* added stage 2 generator
* added stage 2 discriminator
* modified custom layers to accept activation function as a parameter

### CODE REVIEW

**Comment by Devin-Taylor** [comment]
@alecokas keep branch open please

File: models/discriminators.py [hunk 1]
```diff
@@ -126,22 +107,105 @@ def loss(self, predictions_on_real, predictions_on_wrong, predictions_on_fake):
...
+        x = self.conv_block_7(x, training=training)
+
+        res = self.res_block(x, training=training)
+        x = tf.add(x, res)
+        x = tf.nn.leaky_relu(x, alpha=0.2)
+
+        reduced_embedding = self.dense_embed(embedding)
+        reduced_embedding = tf.nn.leaky_relu(reduced_embedding, alpha=0.2)
+        reduced_embedding = tf.expand_dims(tf.expand_dims(reduced_embedding, 1), 1)
+        reduced_embedding = tf.tile(reduced_embedding, [1, 16, 16, 1])
```
Comment (alecokas) [inline models/discriminators.py:193]: I'm a little confused about what this is doing

File: models/discriminators.py [hunk 2]
```diff
@@ -126,22 +107,105 @@ def loss(self, predictions_on_real, predictions_on_wrong, predictions_on_fake):
...
+    def loss(self, predictions_on_real, predictions_on_wrong, predictions_on_fake):
+        """ Calculate the loss for the predictions made on real and fake images.
+                Arguments:
+                predictions_on_real : Tensor
+                predictions_on_fake : Tensor
+        """
+        real_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.ones_like(predictions_on_real), logits=predictions_on_real))
+        wrong_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(predictions_on_wrong), logits=predictions_on_wrong))
+        fake_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(labels=tf.zeros_like(predictions_on_fake), logits=predictions_on_fake))
+        total_loss = real_loss + (wrong_loss + fake_loss) / 2
```
Comment (alecokas) [inline models/discriminators.py:210]: Have we decided that this is the correct way to combine the losses?

File: models/generators.py [hunk 3]
```diff
@@ -127,29 +128,121 @@ def kl_loss(self, mean, log_sigma):
...
+        x = self.conv2d_2(x)
+        x = self.tanh(x)
+
+        return x, mean, log_sigma
+
+    def loss(self, predictions_on_fake, mean, log_sigma):
+        """ Only calculate the loss based on the discriminator
+            predictions for the images generated by this model.
+        """
+        kl_coeff = 2
```
Comment (alecokas) [inline models/generators.py:237]: I guess we also need to make a decision on this one, or try both options?

**Comment by Devin-Taylor** [comment]
Taking a [1, 1, 1, 1] tensor and replicating it to form a [1,16,16,1] vector

**Comment by Devin-Taylor** [comment]
It's what we are doing on stage 1 one just copied for now. But yeah, should
probably make sure we don't need to accumulate losses or something like
that.

**Comment by Devin-Taylor** [comment]
Tried the "reply to this email functionality", clearly failed haha

**Pull request merged**